File: app/src/main/java/eu/rtsketo/sakurastats/main/Interface.java
package eu.rtsketo.sakurastats.main;

import android.content.SharedPreferences;
import android.os.Bundle;
import android.os.SystemClock;
import android.preference.PreferenceManager;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;
import android.support.v4.app.FragmentPagerAdapter;
import android.support.v4.view.ViewPager;
import android.support.v7.app.ActionBar;
import android.support.v7.app.AppCompatActivity;
import android.view.View;
import android.widget.ImageView;
import android.widget.TextView;

import com.qwerjk.better_text.MagicTextView;

import java.util.Timer;
import java.util.TimerTask;

import eu.rtsketo.sakurastats.R;
import eu.rtsketo.sakurastats.control.DataRoom;
import eu.rtsketo.sakurastats.control.DialogView;
import eu.rtsketo.sakurastats.control.ViewDecor;
import eu.rtsketo.sakurastats.fragments.AppSettings;
import eu.rtsketo.sakurastats.fragments.PlayerActivity;
import eu.rtsketo.sakurastats.fragments.Prognostics;
import eu.rtsketo.sakurastats.fragments.WarStatistics;
import eu.rtsketo.sakurastats.hashmaps.PlayerMap;
import eu.rtsketo.sakurastats.hashmaps.SDPMap;

import static eu.rtsketo.sakurastats.control.ThreadPool.getCachePool;
import static eu.rtsketo.sakurastats.control.ViewDecor.decorate;
import static eu.rtsketo.sakurastats.hashmaps.SDPMap.sdp2px;

public class Interface extends AppCompatActivity {
    public static final String TAG = "eu.rtsketo.sakurastats";
    private ImageView progTab, warTab, actiTab, settiTab;
    private MagicTextView[] tab = new MagicTextView[4];
    private SharedPreferences preferences;
    private PlayerActivity actiFrag;
    private WarStatistics warFrag;
    private AppSettings settiFrag;
    private Prognostics progFrag;
    private ViewPager mViewPager;

    private static final int SECS = 1000;
    private static final int MINS = 60*SECS;
    private static final int HRS = 60*MINS;

    public class SectionsPagerAdapter extends FragmentPagerAdapter {
        SectionsPagerAdapter(FragmentManager fm) { super(fm); }
        @Override public int getCount() { return 4; }
        @Override public Fragment getItem(int pos) {
            switch (pos) {
                case 0: return new Prognostics();
                case 1: return new WarStatistics();
                case 2: return new PlayerActivity();
                case 3: return new AppSettings();
                default: return null; }}}

    public void changeTabTo(int num) {
            mViewPager.setCurrentItem(num); }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        SectionsPagerAdapter mSectionsPagerAdapter = new SectionsPagerAdapter(getSupportFragmentManager());
        preferences = PreferenceManager.getDefaultSharedPreferences(getBaseContext());
        setContentView(R.layout.activity_interface);

        mViewPager = findViewById(R.id.viewPager);
        mViewPager.setAdapter(mSectionsPagerAdapter);
        mViewPager.addOnPageChangeListener(pageChangeListener);
        mViewPager.setOffscreenPageLimit(4);

        ActionBar actionBar = getSupportActionBar();
        if (actionBar != null) actionBar.hide();
        ViewDecor.init(getResources());
        SDPMap.init(getResources());
        PlayerMap.init(this);
        DataRoom.init(this);
        initTabs();
        startApp();

        super.onCreate(savedInstanceState);
    }

    @Override
    public void onAttachFragment(Fragment frag) {
        if (frag instanceof PlayerActivity)
            actiFrag = (PlayerActivity) frag;
        else if (frag instanceof WarStatistics)
            warFrag = (WarStatistics) frag;
        else if (frag instanceof AppSettings)
            settiFrag = (AppSettings) frag;
        else if (frag instanceof Prognostics)
            progFrag = (Prognostics) frag;
        super.onAttachFragment(frag);
    }

    private void startApp() {
        getCachePool().execute(() -> {
            if (getLastClan() == null)
                runOnUiThread(() -> new DialogView(
                        DialogView.SakuraDialog.INPUT, this));
            else Service.getThread(this).start(
                    getLastClan(), false, true);
        }); }

    public ImageView getTab(int c) {
        switch (c) {
                case 0: return progTab;
                case 1: return warTab;
                case 2: return actiTab;
                case 3: return settiTab;
                default: throw new
                        NullPointerException(
                                "Invalid number passed in getTab().");
        }

    }

    private void initTabs() {
        progTab = findViewById(R.id.progTab);
        warTab = findViewById(R.id.warTab);
        actiTab = findViewById(R.id.actiTab);
        settiTab = findViewById(R.id.settiTab);

        for (int c = 0; c < 4; c++) {
            final int finalC = c;
            getTab(c).setOnClickListener(
                    view -> changeTabTo(finalC)); }

        tab[0] = findViewById(R.id.tab1);
        tab[1] = findViewById(R.id.tab2);
        tab[2] = findViewById(R.id.tab3);
        tab[3] = findViewById(R.id.tab4);

        int size = sdp2px(7);
        decorate(tab[0], "Forecast", size);
        decorate(tab[1], "Analytics", size);
        decorate(tab[2], "Activity", size);
        decorate(tab[3], "Settings", size);
    }

    public void setLastUse(String tag) { setLastUse(tag, ""); }
    public boolean getLastUse(String tag) { return getLastUse(tag, ""); }
    public void setLastForce(int tab) { setLastUse("tab", getTabName(tab)); }
    public void setLastClan(String tag) {
        SharedPreferences.Editor editor = preferences.edit();
        editor.putString("ClanTag", tag);
        editor.apply(); }

    public String getLastClan() {
        return preferences.getString("ClanTag", getStoredClan(0)); }

    public void setLastUse(String tag, String mod) {
        SharedPreferences.Editor editor = preferences.edit();
        editor.putLong(mod+tag, System.currentTimeMillis());
        editor.apply(); }

    public boolean getLastUse(String tag, String mod) {
        Long curr = System.currentTimeMillis();
        Long time = preferences.getLong(mod+tag, 0);
        long diff = (curr - time);

        switch (mod) {
            case "batt": return diff > 24 * HRS;
            case "prof": return diff > 72 * HRS;
            case "chest": return diff > 24 * HRS;
            case "auto": return diff > 48 * HRS;
            case "wstat": return diff > 48 * HRS;
            case "prog": return diff > 15 * MINS;
            case "acti": return diff > 60 * MINS;
            case "war": return diff > 60 * MINS;
            default: return diff > 15 * MINS; }
    }

    public int getLastForce(int tab) {
        Long curr = System.currentTimeMillis();
        Long time = preferences.getLong(getTabName(tab)+"tab", 0);
        long diff = (curr - time) / 1000 / 60;
        return (int) diff; }

    public int getUseCount() {
        int uc = preferences.getInt("useCount", 0);
        if (uc == 250 || uc == 3000 || uc == 10000)
            runOnUiThread(() -> new DialogView(
                    DialogView.SakuraDialog.RATEQUEST, this));
        return uc; }

    public void incUseCount() {
        SharedPreferences.Editor editor = preferences.edit();
        editor.putInt("useCount", getUseCount()+1);
        editor.apply(); }

    private String getTabName(int tab) {
        switch (tab) {
            case 0: return "prog";
            case 1: return "war";
            default: return "acti"; } }

    public void setStoredClan(int index, String tag) {
        SharedPreferences.Editor editor = preferences.edit();
        editor.putString("StoredClan"+index, tag);
        editor.apply(); }

    public String getStoredClan(int index) {
        return preferences.getString("StoredClan"+index, null); }


    private void changeTab(int num) {
        for (MagicTextView tabetto : tab)
            tabetto.setVisibility(TextView.GONE);
        tab[num].setVisibility(TextView.VISIBLE);
    }

    ViewPager.OnPageChangeListener pageChangeListener = new ViewPager.OnPageChangeListener() {
        @Override public void onPageScrolled(int i, float v, int i1) { }
        @Override public void onPageSelected(int i) { changeTab(i); }
        @Override public void onPageScrollStateChanged(int i) { }
    };

    public void badConnection() {
        runOnUiThread(() -> {
            if (warFrag.getLoading())
                warFrag.getWifi().setVisibility(View.VISIBLE);
            if (actiFrag.getLoading())
                actiFrag.getWifi().setVisibility(View.VISIBLE);
            progFrag.getWifi().first.setVisibility(View.VISIBLE);
            progFrag.getWifi().second.setVisibility(View.VISIBLE);
        });

        Timer timer = new Timer();
        timer.schedule(new TimerTask() {
            @Override public void run() {
                runOnUiThread(() -> {
                    warFrag.getWifi().setVisibility(View.INVISIBLE);
                    actiFrag.getWifi().setVisibility(View.INVISIBLE);
                    progFrag.getWifi().first.setVisibility(View.INVISIBLE);
                    progFrag.getWifi().second.setVisibility(View.INVISIBLE);
                });}},1500); }

    @Override
    public void onBackPressed() {
        this.moveTaskToBack(true);
    }

    public WarStatistics getWarFrag() {
        while (warFrag == null)
            SystemClock.sleep(5);
        waitForView(warFrag);
        return warFrag; }

    public PlayerActivity getActiFrag() {
        while (actiFrag == null)
            SystemClock.sleep(5);
        waitForView(actiFrag);
        return actiFrag; }

    public Prognostics getProgFrag() {
        while (progFrag == null)
            SystemClock.sleep(5);
        waitForView(progFrag);
        return progFrag; }

    public AppSettings getSettiFrag() {
        while (settiFrag == null)
            SystemClock.sleep(5);
        waitForView(settiFrag);
        return settiFrag; }

    private void waitForView(Fragment frag) {
        while (frag.getView() == null)
            SystemClock.sleep(5); }
}

File: app/src/main/java/eu/rtsketo/sakurastats/main/Interface.kt
package eu.rtsketo.sakurastats.main

import android.app.Activity
import android.content.SharedPreferences
import android.os.Bundle
import android.view.View.*
import android.widget.ImageView
import androidx.appcompat.app.AppCompatActivity
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentManager
import androidx.fragment.app.FragmentPagerAdapter
import androidx.viewpager.widget.ViewPager.OnPageChangeListener
import com.qwerjk.better_text.MagicTextView
import eu.rtsketo.sakurastats.R
import eu.rtsketo.sakurastats.control.DataRoom
import eu.rtsketo.sakurastats.control.DialogView
import eu.rtsketo.sakurastats.control.DialogView.SakuraDialog
import eu.rtsketo.sakurastats.control.ThreadPool
import eu.rtsketo.sakurastats.control.ViewDecor
import eu.rtsketo.sakurastats.control.ViewDecor.decorate
import eu.rtsketo.sakurastats.fragments.AppSettings
import eu.rtsketo.sakurastats.fragments.PlayerActivity
import eu.rtsketo.sakurastats.fragments.Prognostics
import eu.rtsketo.sakurastats.fragments.WarStatistics
import eu.rtsketo.sakurastats.hashmaps.PlayerMap
import eu.rtsketo.sakurastats.hashmaps.SDPMap
import eu.rtsketo.sakurastats.hashmaps.SDPMap.Companion.sdp2px
import kotlinx.android.synthetic.main.activity_interface.*
import kotlinx.android.synthetic.main.fragment_activity.*
import kotlinx.android.synthetic.main.fragment_prognose.*
import kotlinx.android.synthetic.main.fragment_prognose.view.*
import kotlinx.android.synthetic.main.fragment_warstats.*
import java.util.*

class Interface : AppCompatActivity() {
    private lateinit var sectionsPagerAdapter: SectionsPagerAdapter
    private lateinit var preferences: SharedPreferences
    private val tab = arrayListOf<MagicTextView>()


    inner class SectionsPagerAdapter internal constructor(fm: FragmentManager)
        : FragmentPagerAdapter(fm, BEHAVIOR_RESUME_ONLY_CURRENT_FRAGMENT) {
        override fun getCount() = 4
        override fun getItem(pos: Int): Fragment {
            return when (pos) {
                0 -> Prognostics.instance
                1 -> WarStatistics.instance
                2 -> PlayerActivity.instance
                else -> AppSettings.instance
            }
        }
    }

    fun changeTabTo(num: Int) {
        viewPager.currentItem = num
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        sectionsPagerAdapter = SectionsPagerAdapter(supportFragmentManager)
        preferences = getPreferences(Activity.MODE_PRIVATE)
        setContentView(R.layout.activity_interface)
        viewPager.adapter = sectionsPagerAdapter
        viewPager.addOnPageChangeListener(pageChangeListener)
        viewPager.offscreenPageLimit = 4
        supportActionBar?.hide()
        ViewDecor.init(resources)
        SDPMap.init(resources)
        PlayerMap.init(this)
        DataRoom.init(this)
        initTabs()
        startApp()
        super.onCreate(savedInstanceState)
    }

//    override fun onAttachFragment(frag: Fragment) {
//        when (frag) {
//            is PlayerActivity -> actiFrag = frag
//            is WarStatistics -> warFrag = frag
//            is AppSettings -> settiFrag = frag
//            is Prognostics -> progFrag = frag
//        }
//        super.onAttachFragment(frag)
//    }

    private fun startApp() {
        ThreadPool.cachePool.execute {
            if (lastClan.isEmpty())
                runOnUiThread {
                    DialogView(SakuraDialog.INPUT, this)
            } else Service.getThread(this)
                    .start(lastClan, force = false, tab = true)
        }
    }

    fun getTab(c: Int): ImageView {
        return when (c) {
            0 -> progTab
            1 -> warTab
            2 -> actiTab
            3 -> settiTab
            else -> throw NullPointerException(
                    "Invalid number passed in getTab().")
        }
    }

    private fun initTabs() {
        for (c in 0..3) getTab(c)
                .setOnClickListener { changeTabTo(c) }
        tab.add(tab1); tab.add(tab2)
        tab.add(tab3); tab.add(tab4)

        val size = sdp2px(7)
        decorate(tab[0], "Forecast", size.toFloat())
        decorate(tab[1], "Analytics", size.toFloat())
        decorate(tab[2], "Activity", size.toFloat())
        decorate(tab[3], "Settings", size.toFloat())
    }

    fun setLastUse(tag: String) {
        setLastUse(tag, "")
    }

    fun getLastUse(tag: String): Boolean {
        return getLastUse(tag, "")
    }

    fun setLastForce(tab: Int) {
        setLastUse("tab", getTabName(tab))
    }

    var lastClan: String
        get() = preferences.getString(
                "ClanTag",
                getStoredClan(0)) ?: ""
        set(tag) {
            val editor
                    = preferences.edit()
            editor.putString("ClanTag", tag)
            editor.apply()
        }

    fun setLastUse(tag: String, mod: String) {
        val editor = preferences.edit()
        editor.putLong(mod + tag, System.currentTimeMillis())
        editor.apply()
    }

    fun getLastUse(tag: String, mod: String): Boolean {
        val curr = System.currentTimeMillis()
        val time = preferences
                .getLong(mod + tag, 0)
        val diff = curr - time
        return when (mod) {
            "batt", "chest" -> diff > 24 * HRS
            "prof" -> diff > 72 * HRS
            "auto", "wstat" -> diff > 48 * HRS
            "prog" -> diff > 15 * MINS
            "acti", "war" -> diff > 60 * MINS
            else -> diff > 15 * MINS
        }
    }

    fun getLastForce(tab: Int): Int {
        val curr = System.currentTimeMillis()
        val time = preferences.getLong(
                getTabName(tab) + "tab", 0)
        val diff = (curr - time) / 1000 / 60
        return diff.toInt()
    }

    private val useCount: Int
        get() {
            val uc = preferences.getInt("useCount", 0)
            if (uc == 20 || uc == 150 || uc == 500) runOnUiThread {
                DialogView(SakuraDialog.RATEQUEST, this) }
            return uc
        }

    fun incUseCount() {
        val editor
                = preferences.edit()
        editor.putInt("useCount", useCount + 1)
        editor.apply()
    }

    private fun getTabName(tab: Int): String {
        return when (tab) {
            0 -> "prog"
            1 -> "war"
            else -> "acti"
        }
    }

    fun setStoredClan(index: Int, tag: String) {
        val editor
                = preferences.edit()
        editor.putString("StoredClan$index", tag)
        editor.apply()
    }

    fun getStoredClan(index: Int): String? {
        val tag = preferences.getString(
                "StoredClan$index", "") ?: ""
        return if (tag.isEmpty()) null else tag
    }

    private fun changeTab(num: Int) {
        for (tabetto in tab)
            tabetto.visibility = GONE
        tab[num].visibility = VISIBLE
    }

    private var pageChangeListener: OnPageChangeListener = object: OnPageChangeListener {
        override fun onPageScrolled(i: Int, v: Float, i1: Int) = Unit
        override fun onPageScrollStateChanged(i: Int) = Unit
        override fun onPageSelected(i: Int) { changeTab(i) }
    }

    fun badConnection() {
            runOnUiThread {
                if (getWarFrag().loading)
                    warFrag?.wifi?.visibility = VISIBLE
                if (getActiFrag().loading)
                    actiFrag?.wifi?.visibility = VISIBLE
                progFrag?.wifi?.visibility = VISIBLE
                progFrag?.wifiOp?.visibility = VISIBLE
            }

            val timer = Timer()
            timer.schedule(object : TimerTask() {
                override fun run() {
                    runOnUiThread {
                        warFrag?.wifi?.visibility = INVISIBLE
                        actiFrag?.wifi?.visibility = INVISIBLE
                        progFrag?.wifi?.visibility = INVISIBLE
                        progFrag?.wifiOp?.visibility = INVISIBLE
                    }
                }
            }, 1500)
    }

    fun getWarFrag()= getFrag(1) as WarStatistics
    fun getActiFrag() = getFrag(2) as PlayerActivity
    fun getProgFrag() = getFrag(0) as Prognostics
    fun getSettiFrag() = getFrag(3) as AppSettings
    private fun getFrag(pos: Int) = sectionsPagerAdapter.getItem(pos)
    override fun onBackPressed() { moveTaskToBack(true) }

    companion object {
        const val TAG = "eu.rtsketo.sakurastats"
        private const val SECS = 1000
        private const val MINS = 60 * SECS
        private const val HRS = 60 * MINS
    }
}